src/main/kotlin/com/pubnub/internal/v2/callbacks/EventEmitterImpl.kt

```diff
@@ -22,6 +22,122 @@ internal class EventEmitterImpl(
 
     private val listeners = CopyOnWriteArraySet<EventListener>()
 
+    @Volatile
```

Comment:
you could greatly simplify the code and lose all the synchronization with this trick:
```
    override var onMessage: ((PubNub, PNMessageResult) -> Unit)? = null
    override var onPresence: ((PubNub, PNPresenceEventResult) -> Unit)? = null
    override var onSignal: ((PubNub, PNSignalResult) -> Unit)? = null
    override var onMessageAction: ((PubNub, PNMessageActionResult) -> Unit)? = null
    override var onObjects: ((PubNub, PNObjectEventResult) -> Unit)? = null
    override var onFile: ((PubNub, PNFileEventResult) -> Unit)? = null

    private val pluggableListener = object : EventListener {
        override fun message(pubnub: PubNub, result: PNMessageResult) {
            onMessage?.invoke(pubnub, result)
        }
        override fun presence(pubnub: PubNub, result: PNPresenceEventResult) {
            onPresence?.invoke(pubnub, result)
        }
        override fun signal(pubnub: PubNub, result: PNSignalResult) {
            onSignal?.invoke(pubnub, result)
        }
        override fun messageAction(pubnub: PubNub, result: PNMessageActionResult) {
            onMessageAction?.invoke(pubnub, result)
        }
        override fun objects(pubnub: PubNub, result: PNObjectEventResult) {
            onObjects?.invoke(pubnub, result)
        }
        override fun file(pubnub: PubNub, result: PNFileEventResult) {
            onFile?.invoke(pubnub, result)
        }
    }.apply {
        addListener(this)
    }
```

I haven't checked if this works though

Reply:
Please, have a look at changes. They are kind of in line with what you proposed but not exactly.

Reply:
I don't really like the proposed solution for the following reasons:
- it still needs synchronization
- it's much longer and has more complex code
- it adds casting

Are there any advantages to the solution you proposed vs the single listener approach in my comment above, or reasons why my solution wouldn't work?